### DISCUSSION TITLE
Persist storage : values are converted to strings before replacer

### DISCUSSION THREAD

**Discussion by karbone4** [pmndrs/zustand#2403] [Bug report]
## Summary

I am trying to use `persist` with dates.

Bug: values are converted to strings before replacer function so I cannot convert it back to date because I couldn't apply `type: 'date'` to convert it to date in `reviver`

## Link to reproduction

[Code sandbox](https://codesandbox.io/p/sandbox/bug-zustand-replacer-hjg9xw?file=%2Fsrc%2FApp.js%3A21%2C23)

![image](https://github.com/pmndrs/zustand/assets/48047952/62772854-76d9-45cc-a55a-65facb64eb8c)

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Do you have an idea how to fix it?

**Reply by karbone4** [reply]
I followed the docs https://docs.pmnd.rs/zustand/integrations/persisting-store-data#createjsonstorage and it said it works so I am doing something wrong or the doc is wrong ?

**Reply by dai-shi** [reply]
reviver/replacer options are added relatively recently, so I'm not sure how many people use it. Shouldn't be too many. So, it's not too surprising if there was a bug.

It's developed in #1763 and the docs was added in #2299.
Maybe @lauhon or @charkour could help.

**Reply by karbone4** [reply]
I use it to convert back dates from my storage. I update code sandbox with examples. I want to do it with dates and nested dates :
```
{
      testDate: new Date(),
      testNestedDate: {
        date: new Date(),
      },
    }
```
I think reviver / replacer only uses the `JSON.parse` / `JSON.stringfy` of the entire object store in zustand

**Reply by charkour** [reply]
Hi @karbone4, could you clarify how to use the code sandbox to reproduce the bug? The `replacer` you set is passed to `JSON.stringify` so I don't think it should convert your value to a string _before_ the `replacer` runs.

Here is the source code: https://github.com/pmndrs/zustand/blob/c092d76260fb752eb6b7c735278c083e7227a805/src/middleware/persist.ts#L59

**Reply by karbone4** [reply]
Thanks for additional informations, I modified the code sandbox with new reviver and replacer that works
```
{
        reviver: (key, value: any) => {
          if (value && key === "date") {
            return parseISO(value.value);
          }
          return value;
        },
        replacer: (key, value) => {
          if (value && key === "date") {
            return {
              type: "date",
              value: value instanceof Date ? value.toISOString() : value,
            };
          }
          return value;
        },
      }
```
It handles the fact that date values are converted to string by JSON.stringfy.

**Reply by dai-shi** [reply]
Nice. Feel free to submit a PR to improve the docs.

**Reply by coldiary** [reply]
This would not be a suitable solution for every case, as it depends on a specific key to determine that the value is a date.

I've come to found out that _if using a function for the replacer to keep the context, the value of `this[key]` remains the original value while the `value` arg is the serialized one. For the reviver, `this[key]` is undefined, and `value` is the serialized one._

So we used something like this, for it to work : 

```js
  {
    replacer: function (key) {
      const v = this[key];
      
      if (v instanceof Date) {
        return { __type: 'Date', value: v.toISOString() };
      }

      return v;
    },
    reviver: function(_, value) {
      if (value?.__type === 'Date') {
        return parseISO(value.value);
      }

      return value;
    },
  }
```

**Comment by dai-shi** [comment]
Here's the test code:
https://github.com/pmndrs/zustand/blob/c092d76260fb752eb6b7c735278c083e7227a805/tests/persistSync.test.tsx#L698-L708

Hmm, I'm not sure how it's not working with `Date`.

**Comment by quertc** [comment]
The same problem with Duration / Timestamp  from @bufbuild/protobuf.

**Comment by 60pfennig** [comment]
I have the same issue. Its also problematic that in the docs exactly this bug is written as [an example](https://zustand.docs.pmnd.rs/integrations/persisting-store-data#createjsonstorage)
This has lead to an enourmous debug session.

**Reply by 60pfennig** [reply]
And wouldn't it be better to keep the conversation in the issue instead of closing it? In a discussion its harder to find and for me this seems like a potential bug and not a request for help.
